src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8118,29 +8133,63 @@ void checkWellKnown(WellKnownMember member)
 
             var lookupResult = LookupResult.GetInstance();
 
-            if (TryLookupLengthOrCount(receiverType, lookupResult, out lengthOrCountProperty, ref useSiteInfo) &&
-                TryFindIndexOrRangeIndexerPattern(lookupResult, receiverOpt, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo))
+            if (TryBindLengthOrCount(syntax, receiverOpt, receiverType, out lengthOrCountAccess, diagnostics) &&
+                TryFindIndexOrRangeImplicitIndexer(syntax, lookupResult, receiverOpt, receiverType, argIsIndex, out implicitIndexerAccess, out argumentPlaceholders, diagnostics))
+            {
+                CheckValue(lengthOrCountAccess, BindValueKind.RValue, diagnostics);
+
+                lookupResult.Free();
+                return true;
+            }
+
+            lengthOrCountAccess = null;
+            implicitIndexerAccess = null;
+            argumentPlaceholders = default;
+            lookupResult.Free();
+            return false;
+        }
+
+        private bool TryBindLengthOrCount(
+            SyntaxNode syntax,
+            BoundExpression? receiverOpt,
+            TypeSymbol receiverType,
+            [NotNullWhen(true)] out BoundExpression? lengthOrCountAccess,
+            BindingDiagnosticBag diagnostics)
+        {
+            var lookupResult = LookupResult.GetInstance();
+
+            if (TryLookupLengthOrCount(syntax, receiverType, lookupResult, out var lengthOrCountProperty, diagnostics))
             {
-                CheckImplicitThisCopyInReadOnlyMember(receiverOpt, lengthOrCountProperty.GetMethod, diagnostics);
+                lengthOrCountAccess = BindPropertyAccess(syntax, receiverOpt, lengthOrCountProperty, diagnostics, lookupResult.Kind, hasErrors: false);
+                ReportDiagnosticsIfObsolete(diagnostics, lengthOrCountProperty, syntax, hasBaseReceiver: false);
+
                 lookupResult.Free();
                 return true;
             }
 
-            patternSymbol = null;
+            lengthOrCountAccess = null;
             lookupResult.Free();
+
             return false;
         }
 
-        private bool TryFindIndexOrRangeIndexerPattern(
+        /// <summary>
+        /// Finds pattern-based implicit indexer:
+        /// - for Index indexer, this will find `this[int]`.
+        /// - for Range indexer, this will find `Slice(int, int)` or `string.Substring(int, int)`.
+        /// </summary>
+        private bool TryFindIndexOrRangeImplicitIndexer(
```

Comment:
>TryFindIndexOrRangeImplicitIndexer [](http://example.com/codeflow?start=21&length=34)

I think the name and the comment for the method is confusing:
1. It is not about an implicit indexer, but about a real indexer or a slice method that is used to implement an implicit indexer by the compiler.
2. It binds that underlying access, rather than finds it.